### DISCUSSION TITLE
Memcpy issue on PicoW

### DISCUSSION THREAD

**Discussion by buhlerles** [earlephilhower/arduino-pico#2482] [Q&A]
Having an issue with memcpy on a Pico W in Arduino with Earl's core that is most puzzling (Arduino 1.8...)

memcpy appears to put the array value at the end of the segement to copy into ALL the memory locations

Code to reproduce:
     uint8_t test[11] = {0, 0, 0, 0, 0, 0, 0, 0, 0, 67, 0};
      // Shift the last nine bytes one byte left.
      memcpy(test, test + 1, 9);
      // memcpy is putting the last value in all the array elements ??? !!!
      for ( uint8_t i = 0; i < (11); i++)
      {
        Serial.print(test[i]);
        Serial.print("  ");
      }
      Serial.println();

I'm anything but an expert and may be overlooking something totally basic, but the same code runs fine on a Teensy and does what you'd expect.

Any insights would be most appreciated!

**Comment by earlephilhower** [comment]
`memcpy` is illegal/undefined when the source and destination overlap.  Use `memmove`.

https://man7.org/linux/man-pages/man3/memcpy.3.html

**Comment by buhlerles** [comment]
Aha - was not aware of that limitation!
Thank-you!
